governance/program/src/state/proposal.rs

```diff
@@ -385,48 +390,96 @@ impl ProposalV2 {
                     return Ok(governing_token_mint_supply);
                 }
 
-                let max_vote_weight = (governing_token_mint_supply as u128)
+                let max_voter_weight = (governing_token_mint_supply as u128)
                     .checked_mul(fraction as u128)
                     .unwrap()
                     .checked_div(MintMaxVoteWeightSource::SUPPLY_FRACTION_BASE as u128)
                     .unwrap() as u64;
 
-                let deny_vote_weight = self.deny_vote_weight.unwrap_or(0);
-
-                let max_option_vote_weight =
-                    self.options.iter().map(|o| o.vote_weight).max().unwrap();
-
                 // When the fraction is used it's possible we can go over the calculated max_vote_weight
                 // and we have to adjust it in case more votes have been cast
-                let total_vote_weight = max_option_vote_weight
-                    .checked_add(deny_vote_weight)
-                    .unwrap();
-
-                Ok(max_vote_weight.max(total_vote_weight))
+                Ok(self.coerce_max_voter_weight(max_voter_weight))
             }
             MintMaxVoteWeightSource::Absolute(_) => {
                 Err(GovernanceError::VoteWeightSourceNotSupported.into())
             }
         }
     }
 
+    /// Adjusts max voter weight to ensure it's not lower than total cast votes
+    fn coerce_max_voter_weight(&self, max_voter_weight: u64) -> u64 {
+        let deny_vote_weight = self.deny_vote_weight.unwrap_or(0);
+
+        let max_option_vote_weight = self.options.iter().map(|o| o.vote_weight).max().unwrap();
+
+        let total_vote_weight = max_option_vote_weight
+            .checked_add(deny_vote_weight)
+            .unwrap();
+
+        max_voter_weight.max(total_vote_weight)
+    }
+
+    /// Resolves max voter weight
+    pub fn resolve_max_voter_weight(
+        &mut self,
+        program_id: &Pubkey,
+        realm_config_info: &AccountInfo,
+        governing_token_mint_info: &AccountInfo,
+        account_info_iter: &mut Iter<AccountInfo>,
+        realm: &Pubkey,
+        realm_data: &Realm,
+    ) -> Result<u64, ProgramError> {
+        // if the realm uses addin for max community voter weight then use the externally provided max weight
+        if realm_data.config.use_max_community_voter_weight_addin
+            && realm_data.community_mint == self.governing_token_mint
+        {
+            let realm_config_data =
+                get_realm_config_data_for_realm(program_id, realm_config_info, realm)?;
+
+            let max_voter_weight_record_info = next_account_info(account_info_iter)?;
+
+            let max_voter_weight_data =
```

Comment:
Looking at this makes me realize that these addins are huge attack vectors!  I hope they have ample protections in place to be sure that no one else can create records on an existing program.  From what I see, there's no check that the max voter weight record address corresponds to any PDA or specific address.  Is that correct?

Reply:
You are right, `MaxVoterWeightRecord` and `VoterWeightRecord` accounts belong to the `addin` and not to the `spl-governance` and hence the onus of protecting them is on the addin side because `spl-governance` fully trust  the data provided via the accounts and It only verifies the integrity of the data